docs/book/reflection-abstract-factory.md

```diff
@@ -6,10 +6,40 @@ Writing a factory class for each and every service that has dependencies
 can be tedious, particularly in early development as you are still sorting
 out dependencies.
 
-laminas-servicemanager ships with `Laminas\ServiceManager\AbstractFactory\ReflectionBasedAbstractFactory`,
-which provides a reflection-based approach to instantiation, resolving
-constructor dependencies to the relevant services. The factory may be used as
-either an abstract factory, or mapped to specific service names as a factory:
+To alleviate this issue during development, laminas-servicemanager ships with the [ReflectionBasedAbstractFactory](https://github.com/laminas/laminas-servicemanager/blob/3.12.x/src/AbstractFactory/ReflectionBasedAbstractFactory.php),
+which provides a [reflection-based approach](https://www.php.net/manual/en/intro.reflection.php) to instantiation, 
+resolving constructor dependencies to the relevant services. 
+The factory may be used as either an abstract factory or mapped to specific service names as a factory.
+
+TIP: Mapping services to the factory is more explicit and performant.
+
+The factory operates with the following constraints/features:
+
+- A parameter named `$config` type-hinted as an array will receive the
+  application "config" service (i.e., the merged configuration).
+- Parameters type-hinted against array, but not named `$config`, will
+  be injected with an empty array.
```

Comment:
Urgh, we do this? This is terrible :O

Reply:
Yes, since [version 3.0.1 of zend-mvc](https://github.com/zendframework/zend-mvc/blob/master/CHANGELOG.md#301---2016-06-23) (2016-06-23).
And this is also documented: 

> The factory operates with the following constraints/features:
> 
> - A parameter named $config typehinted as an array will receive the application "config" service (i.e., the merged configuration).
> - Parameters typehinted against array, but not named $config, will be injected with an empty array.
> - Scalar parameters will result in the factory raising an exception, unless a default value is present; if it is, that value will be used.
> - If a service cannot be found for a given typehint, the factory will raise an exception detailing this.

https://docs.laminas.dev/laminas-servicemanager/reflection-abstract-factory/#reflection-factory